### DISCUSSION TITLE
Add/extract feature-level metadata in a Seurat v5 object?

### DISCUSSION THREAD

**Discussion by nick-youngblut** [satijalab/seurat#8993] [Q&A]
I have a data.frame of metadata for all genes in my Seurat v5 object. 

The [Seurat Command List docs](https://satijalab.org/seurat/articles/essential_commands) include a section on [cell metadata](https://satijalab.org/seurat/articles/essential_commands#cell-metadata), but no such section on feature metadata. 

How does one add a data.frame of feature-level metadata to a Seurat v5 object (preferably matching feature names by the rownames in the metadata data.frame)? 

How does one then extract the table of the feature-level metadata from the Seurat v5 object?

**Comment by wudustan** [comment]
Feature metadata isn't currently supported. Seurat objects have a `misc` slot you can access with `Misc(x)` and can have values added to it with `Misc(x) <- some.data`

More info can be found by running `?SeuratObject::Misc`

**Comment by nick-youngblut** [comment]
> Feature metadata isn't currently supported

I am currently storing feature metadata via `seurat_obj@assays$RNA@meta.data`. 

I'm surprised that feature metadata isn't supported, given that valuable metadata is often at the feature-level (e.g., Ensemble accessions corresponding to gene names). For instance, how does one deal with possible alternative namings for features (e.g., your reference dataset uses Ensembl accessions, while your query dataset uses gene names)? 

@wudustan how do you deal with feature metadata or renaming features for the purpose of matching features across datasets?

**Reply by wudustan** [reply]
Tools like Azimuth do feature matching and conversion on the fly. Other tools assume that you're happy to handle that kind of stuff manually, and sometimes you might want to retain more control to make sure you know what's being done.

I have recently had to integrate a large dataset coming from lots of different sources and I manually integrated features at the point of creating individual Seurat objects with just counts matrices then reducing on features I wanted to keep, making sure that some features weren't only unique to one dataset etc. By the time I had a merged Seurat they were all lined up. 

I suppose if I had to keep additional feature information for this purpose I would store it as some kind of data.frame/list object. Not ideal but as long as you're consistent it's probably enough - I do agree that it would be nice to have officual support for this and you could add a feature request in the issues. 

If you're keen enough you could try implementing something by modifying the Seurat object definition and the devs will probably be happy to integrate this.

**Comment by nick-youngblut** [comment]
> I manually integrated features at the point of creating individual Seurat objects with just counts matrices then reducing on features I wanted to keep, making sure that some features weren't only unique to one dataset etc. By the time I had a merged Seurat they were all lined up

Sounds like a lot of work... which could be automated.

> I do agree that it would be nice to have officual support for this and you could add a feature request in the issue

👍 

> If you're keen enough you could try implementing something by modifying the Seurat object definition and the devs will probably be happy to integrate this

I'm fairly new to Seurat, but not new to R development, so I might give it a shot.

Thanks for all of the info!
